text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,174 @@
+- Feature Name: macro-at-most-once-rep
+- Start Date: 2018-01-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+
+Summary
+-------
+
+Add a repetition specifier to macros to repeat a pattern at most once: `$(pat)?`. Here, `?` behaves like `+` or `*` but represents at most one repetition of `pat`.
+
+Motivation
+----------
+
+There are two specific use cases in mind.
+
+## Macro rules with optional parts
+
+Currently, you just have to write two rules and possibly have one "desugar" to the other.
+
+```rust
+macro_rules! foo {
+  (do $b:block) => {
+    $b
+  }
+  (do $b1:block and $b2:block) => {
+    foo!($b1)
+    $b2
+  }
+}
+```
+
+Under this RFC, one would simply write:
+
+```rust
+macro_rules! foo {
+  (do $b1:block $(and $b2:block)?) => {
+    $b1
+    $($b2)?
+  }
+}
+```
+
+## Trailing commas
+
+Currently, the best way to make a rule tolerate trailing commas is to create another identical rule that has a comma at the end:
+
+```rust
+macro_rules! foo {
+  ($(pat),*,) => { foo!( $(pat),* ) };
```

Comment:
This rule should be `$(pat),+,` so that `foo!(,)` is not permitted.

Reply:
:+1: will fix shortly